File: leetcode/LC/linkedlist/0025_reverse_nodes_in_k_group.cpp
/**
 * Definition for singly-linked list.

 */
#include <iostream>

  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

// https://www.youtube.com/watch?v=LCRGV8avvUY

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int count = 0;
        ListNode* temp = head;
        while(temp && count <k){
            temp = temp->next;
            count++;
        }
        
        if(count == k){
            count = 0;
            temp = head;
            ListNode* prev= NULL;
            ListNode* next = NULL;
            while(temp && count < k){
                next = temp->next;
                temp->next = prev;
                prev = temp;
                temp = next;
                count++;
            }
            head->next = reverseKGroup(next, k);
            if(!prev)prev=head;
            return prev;
        }else
            return head;
    }
};

File: leetcode/LC1/linkedlist/0025_reverse_nodes_in_k_group.py
from typing import Optional

# https://www.youtube.com/watch?v=LCRGV8avvUY

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        n = 0
        prev = None
        current = head
        
        length = 0
        temp = head
        while temp != None:
            length += 1
            temp = temp.next
        
        l = length // k
        
        while(n < l):
            m = 0
            dummy = current
            dummy1 = prev
            while m <k:
                temp = current.next
                current.next = prev
                prev = current
                current = temp
                m += 1
            if (n == 0):
                head = prev
            dummy.next = current
            if dummy1:
                dummy1.next = prev
            prev = dummy
            n += 1
        return head
        